### PR TITLE
[#9] [Integrate] Login screen

### PR DESCRIPTION
**Pull request by AVI5HEK**
https://github.com/nimblehq/ic-flutter-avishek/issues/9

## What happened 👀

- Validate email and password (email must have a correct format, password must be filled).
- Call the login endpoint which was done in https://github.com/nimblehq/ic-flutter-avishek/issues/8 when user tap on Log in button.
- Save the retrieved cridentials for later use.

## Insight 📝

- Fixed an issue with JSON parsing using the JAPX package.
- Used the stored access token (implemented in #8) in the interceptor.
- Added UI tests.

## Proof Of Work 📹

https://github.com/nimblehq/ic-flutter-avishek/assets/8093908/7e0cad22-5622-443e-92e6-dbb5c526a938

### CODE REVIEW

File: lib/ui/widget/login_form.dart [hunk 1]
```diff
@@ -100,12 +115,41 @@ class LoginFormState extends ConsumerState<LoginForm> {
...
-            },
+            onPressed: () => {_logIn()},
           ),
         ),
       ],
     );
   }
+
+  String? _emailValidator(String? value) {
+    if (value == null || !EmailValidator.validate(value)) {
```
Comment (markgravity) [inline lib/ui/widget/login_form.dart:126]: What do you think about moving this logic to the `LoginViewModel`? 🙏

File: lib/ui/widget/login_form.dart [hunk 2]
```diff
@@ -100,12 +115,41 @@ class LoginFormState extends ConsumerState<LoginForm> {
...
+      return AppLocalizations.of(context)!.errorInvalidEmail;
+    } else {
+      return null;
+    }
+  }
+
+  String? _passwordValidator(String? value) {
+    if (value == null || value.length < _minPasswordLength) {
```
Comment (markgravity) [inline lib/ui/widget/login_form.dart:134]: Same here

File: lib/ui/widget/login_form.dart (hunk 1 above)
Comment (AVI5HEK) [reply lib/ui/widget/login_form.dart]: Resolved in https://github.com/nimblehq/ic-flutter-avishek/pull/36/commits/87733e49c74585d40d26767b5e685a67875506a2

File: lib/ui/widget/login_form.dart (hunk 2 above)
Comment (AVI5HEK) [reply lib/ui/widget/login_form.dart]: Resolved in https://github.com/nimblehq/ic-flutter-avishek/pull/36/commits/87733e49c74585d40d26767b5e685a67875506a2

**Review by markgravity — APPROVED** [review]

File: lib/model/request/auth_token_request.dart [hunk 3]
```diff
@@ -11,9 +11,9 @@ const _grantTypePassword = "password";
 class AuthTokenRequest {
   final String email;
   final String password;
-  final String grantType = _grantTypePassword;
-  final String clientId = Env.authClientId;
-  final String clientSecret = Env.authClientSecret;
+  late String grantType = _grantTypePassword;
+  late String clientId = Env.authClientId;
+  late String clientSecret = Env.authClientSecret;
```
Comment (Wadeewee) [inline lib/model/request/auth_token_request.dart:16]: @AVI5HEK Why not we call them directly in `AuthenticationRepositoryImpl` so we can get rid of defining `late` 🤔

**Review by Wadeewee — COMMENTED** [review]
For the rest LGTM 🚀

**Review by Wadeewee — APPROVED** [review]

**Pull request merged**